core/src/main/java/org/apache/accumulo/core/iterators/user/CfCqSliceSeekingFilter.java

```diff
@@ -0,0 +1,132 @@
+  private static final FilterResult SKIP_TO_NEXT_ROW = new FilterResult(false, AdvanceResult.NEXT_ROW);
+  private static final FilterResult SKIP_TO_NEXT_CF = new FilterResult(false, AdvanceResult.NEXT_CF);
+  private static final FilterResult INCLUDE_AND_NEXT = new FilterResult(true, AdvanceResult.NEXT);
+  private static final FilterResult INCLUDE_AND_NEXT_CF = new FilterResult(true, AdvanceResult.NEXT_CF);
+
+  private CfCqSliceOpts cso;
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+    super.init(source, options, env);
+    cso = new CfCqSliceOpts(options);
+  }
+
+  @Override
+  public FilterResult filter(Key k, Value v) {
+    if (cso.minCf.getLength() > 0) {
+      int minCfCmp = k.compareColumnFamily(cso.minCf);
+      if (minCfCmp < 0) {
+        return SKIP_TO_HINT; // hint will be the min CF in this row.
+      }
+      if (minCfCmp == 0 && !cso.minInclusive) {
+        return SKIP_TO_NEXT;
+      }
+    }
+    if (cso.maxCf.getLength() > 0) {
+      int maxCfCmp = k.compareColumnFamily(cso.maxCf);
+      if (maxCfCmp > 0 || (maxCfCmp == 0 && !cso.maxInclusive)) {
+        return SKIP_TO_NEXT_ROW;
+      }
+    }
+    // at this point we're in the correct CF range, now check the CQ.
+    if (cso.minCq.getLength() > 0) {
+      int minCqCmp = k.compareColumnQualifier(cso.minCq);
+      if (minCqCmp < 0) {
+        return SKIP_TO_HINT; // hint will be the min CQ in this CF in this row.
+      }
+      if (minCqCmp == 0 && !cso.minInclusive) {
+        return SKIP_TO_NEXT;
+      }
+    }
+    if (cso.maxCq.getLength() > 0) {
+      int maxCqCmp = k.compareColumnQualifier(cso.maxCq);
+      if (maxCqCmp > 0 || (maxCqCmp == 0 && !cso.maxInclusive)) {
+        return SKIP_TO_NEXT_CF;
+      }
+      if (maxCqCmp == 0) {
+        // special-case here: we know we're at the last CQ in the slice, so skip to the next CF in the row.
+        return INCLUDE_AND_NEXT_CF;
+      }
+    }
+    // at this point we're in the CQ slice.
+    return INCLUDE_AND_NEXT;
+  }
+
+  @Override
+  public Key getNextKeyHint(Key k, Value v) {
+    if (cso.minCf.getLength() > 0) {
+      int minCfCmp = k.compareColumnFamily(cso.minCf);
+      if (minCfCmp < 0) {
+        Key hint = new Key(k.getRow(), cso.minCf);
+        return cso.minInclusive ? hint : hint.followingKey(PartialKey.ROW_COLFAM);
+      }
+    }
+    if (cso.minCq.getLength() > 0) {
+      int minCqCmp = k.compareColumnQualifier(cso.minCq);
+      if (minCqCmp < 0) {
+        Key hint = new Key(k.getRow(), k.getColumnFamily(), cso.minCq);
+        return cso.minInclusive ? hint : hint.followingKey(PartialKey.ROW_COLFAM_COLQUAL);
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+    CfCqSliceSeekingFilter o = (CfCqSliceSeekingFilter) super.deepCopy(env);
```

Comment:
I noticed deepCopy was not covered by unit test.   This would be easier to test in the branch I created that replaced MAC with SortedMapIterator in the unit test.